### PR TITLE
ball_filter integration

### PR DESCRIPTION
**Pull request by NFiedler**
## Proposed changes
Integrate ball filter into behavior

## Related issues
#131 
This only works in combination with https://github.com/bit-bots/bitbots_world_model/pull/21

## Necessary checks
- [ ] Update package version
- [ ] Run `catkin build`
- [ ] Write documentation
- [ ] Create issues for future work
- [ ] Test on your machine
- [ ] Test on the robot
- [ ] Put the PR on our Project board

### CODE REVIEW

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py [hunk 1]
```diff
@@ -207,35 +208,30 @@ def get_ball_angle(self):
...
-                self.ball_seen = True
+    def ball_filtered_callback(self, msg: PoseWithCovarianceStamped):
+        self.ball_filtered = msg
 
-            except (tf2.ConnectivityException, tf2.LookupException, tf2.ExtrapolationException) as e:
-                rospy.logwarn(e)
+        # When the precision is not sufficient, the ball ages.
+        x_sdev = msg.pose.covariance[0]  # position 0,0 in a 6x6-matrix
+        y_sdev = msg.pose.covariance[7]  # position 1,1 in a 6x6-matrix
+        if x_sdev > self.body_config['ball_position_precision_threshold']['x_sdev'] or \
```
Comment (Flova) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:217]: The covariance gets checked for both the position and the twist. Both times the cov in x and y direction are used. Why is there a different parameter for them?

File: bitbots_body_behavior/config/body_behavior.yaml [hunk 2]
```diff
@@ -138,7 +143,7 @@ behavior:
     ball_reaproach_dist: 1.0
 
     # Distance at which the ball is normally approached
-    ball_approach_dist: 0.2
+    ball_approach_dist: 0.25
```
Comment (Flova) [inline bitbots_body_behavior/config/body_behavior.yaml:146]: Why is this param changed? This may break things significantly, because it is out of the kick area. This results in the behavior from the game against Starkit. Also, why is this necessary? Is the new ball position less accurate and we, therefore, need more distance to avoid it while kicking?

File: bitbots_body_behavior/src/bitbots_body_behavior/body_behavior.py [hunk 3]
```diff
@@ -56,7 +56,7 @@
     D.blackboard.dynup_action_client = actionlib.SimpleActionClient('dynup', DynUpAction)
 
     # TODO: callbacks away from the blackboard!
-    rospy.Subscriber("balls_relative", PoseWithCertaintyArray, D.blackboard.world_model.balls_callback)
+    # rospy.Subscriber("balls_relative", PoseWithCertaintyArray, D.blackboard.world_model.balls_callback)
```
Comment (Flova) [inline bitbots_body_behavior/src/bitbots_body_behavior/body_behavior.py:59]: Please remove this line if it is unused.

**Review by Flova — CHANGES_REQUESTED** [review]

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py [hunk 4]
```diff
@@ -43,8 +43,8 @@ def to_pose_with_certainty_array(self):
 class WorldModelCapsule:
     def __init__(self, blackboard):
         self._blackboard = blackboard
+        self.body_config = rospy.get_param("behavior/body")
```
[the remark this answers is not in the document]
Comment (NFiedler) [reply bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:46]: it is not the same when it is used in the head behavior.

[the remark this answers is not in the document]
Comment (Flova) [reply bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:46]: But then the body params aren't loaded, because they are not loaded by the head behavior.

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py (hunk 1 above)
Comment (NFiedler) [reply bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:217]: to allow more flexibility for example in a case of filtering in the map frame. This does not restrict us in any way.

File: bitbots_body_behavior/config/body_behavior.yaml (hunk 2 above)
Comment (NFiedler) [reply bitbots_body_behavior/config/body_behavior.yaml:146]: ill change it back

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NFiedler) [reply bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:46]: Well, this solved an issue which crashes the head behavior. We could load the body behavior config when the head behavior is started without a body behavior.

File: bitbots_body_behavior/src/bitbots_body_behavior/body_behavior.py (hunk 3 above)
Comment (NFiedler) [reply bitbots_body_behavior/src/bitbots_body_behavior/body_behavior.py]: done.

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Flova) [reply bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:46]: We should add the config param to the head behavior config and use the `self.blackboard.config`. The head behavior features also params for the ball confidence at the moment afaik.

[the remark this answers is not in the document]
Comment (NFiedler) [reply bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:46]: thats ugly as we could have different parameters between head and body behavior :(

**Review by Flova — APPROVED** [review]

**Review by SammyRamone — APPROVED** [review]
lgtm, but I did not test it yet
